### DISCUSSION TITLE
Copilot in IntelliJ IDEA crashed when logging in to server

### DISCUSSION THREAD

**Discussion by lyh543** [community/community#16230] [Copilot Conversations] [answered]
I'm using IntelliJ **IDEA 2022.1#IU-221.5080.210** and **GitHub Copilot 1.1.20.1417** on **Windows**.

When I click `Login to GitHub`, IDEA always throws Exception:

```
JsonRpcErrorException(requestId=3, code=-32603, extraData=null)
	at com.github.copilot.lang.agent.rpc.JsonRPC.parseResponse(JsonRPC.java:52)
	at com.github.copilot.lang.agent.vscodeRpc.DefaultJsonRpcMessageHandler.handleJsonMessage(DefaultJsonRpcMessageHandler.java:74)
	at com.github.copilot.lang.agent.vscodeRpc.VSCodeJsonRpcParser.processPendingContent(VSCodeJsonRpcParser.java:91)
	at com.github.copilot.lang.agent.vscodeRpc.VSCodeJsonRpcParser.appendLocked(VSCodeJsonRpcParser.java:71)
	at com.github.copilot.lang.agent.vscodeRpc.VSCodeJsonRpcParser.append(VSCodeJsonRpcParser.java:39)
	at com.github.copilot.lang.agent.vscodeRpc.AgentProcessHandler$1.onTextAvailable(AgentProcessHandler.java:35)
	at jdk.internal.reflect.GeneratedMethodAccessor98.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.intellij.execution.process.ProcessHandler$2.invoke(ProcessHandler.java:248)
        .......
	at java.base/java.lang.Thread.run(Thread.java:829)
```

I downgraded to multiple Copilot versions, and the tests results are below:

|Copilot  version|works?|
|-|-|
|[1.1.20.1450-agent](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/nightly/173803)|no|
|[1.1.20.1417](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/172765)|no|
|[1.1.19.1364](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/171146)|no|
|[1.1.17.1262](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/169432)|no|
|[1.1.16.1219](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/167546)|was, but too old to connect to GitHub Server after 2022.5.20|
|[1.1.12.524](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/158042)|was, but too old to connect to GitHub Server after 2022.5.11|

I'd like to do more tests if needed.

--------------------

**2022.5.23 Workaround if you are able to connect to GitHub without proxy**

1. start a terminal
2. run `export HTTP_PROXY=""; export HTTPS_PROXY=""` (for Mac OS or Linux) or `$env:HTTP_PROXY = $env:HTTPS_PROXY = ""` (for PowerShell on Windows)
3. in the same terminal, run `/path/to/jetbrains/idea`
4. to validate `HTTP_PROXY` is clear in JetBrains, you can run a terminal in JetBrain, and `echo $HTTP_PROXY` or `echo $env:HTTP_PROXY`

--------------------

**2022.5.20 Workaround if you MUST use proxy to connect to GitHub (success on Windows, other OS may work with another HTTPS package capturing tool)**

1. run your original proxy on `http://localhost:8080` (or any other port) and set as Windows system proxy.
2. run Fiddler Classic (a web captruing tool) and listen on `http://localhost:8888` (Fiddler Classic -> Tools -> Options -> Connections)
3. set Fiddler Classic -> Tools -> Options -> HTTPS as belows
![image](https://user-images.githubusercontent.com/15522311/169429393-b04fe74f-dadb-41c8-ad01-cb6f78ce94f0.png)
4. start a terminal
5. run `export HTTP_PROXY="http://localhost:8888"; export HTTPS_PROXY="http://localhost:8888"` (for Mac OS or Linux) or `$env:HTTP_PROXY = $env:HTTPS_PROXY = "http://localhost:8888"` (for PowerShell on Windows)
6. in the same terminal, run `/path/to/jetbrains/idea`
7. in IDEA, trust all certificates named `DO_NOT_TRUST_FiddlerRoot`, and now Copilot should work with proxy.

I guess the point is that `axios` uses the TLS version that is incompatible with my proxy. Fiddler decrypts the HTTPS package and then re-encrypts with compatiable TLS version. 

This workaround has two downsides:

1. Fiddler Classic should keep running
2. All packages from applications using system proxy or `HTTP_PROXY` env will be decrypt by Filddler, and you need to trust Fiddler certificate anywhere. (This can be fixed by another workaround: set system proxy and `HTTP_PROXY` and `HTTPS_PROXY` as `http://localhost:8080`, and start IntelliJ in a Powershell or cmd with `HTTP_PROXY=HTTPS_PROXY=http://localhost:8888`)

**Comment by lyh543** [comment]
Now I downgrade Copilot to [1.1.16.1219](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/167546), and it works.

**Reply by zukxu** [reply]
when I login to github ,the web page is successed, but the IDEA is always in loading ,that show  Waiting for Github Authentication

could you help me resolve this problem
![image](https://user-images.githubusercontent.com/63381545/167465898-20ebcdc3-746d-4d49-803e-816c480f791d.png)

![image](https://user-images.githubusercontent.com/63381545/167465863-e1b7be73-f5bf-4e03-9574-234f3e08a827.png)

**Reply by lyh543** [reply]
@zukxu Set proxy for IDEA :)

**Reply by zukxu** [reply]
I set up a global proxy for my computer, with Clash? but still not. Do I need to set a proxy for IDEA separately?

**Reply by lyh543** [reply]
@zukxu Some softwares read and use system global proxy settings, some use `$env:HTTP_PROXY`, and some have own configures. JetBrains softwares are the last ones.

**Reply by ihx-rainbow** [reply]
@zukxu [GitHub Copilot 1.1.20.1450-agent](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/nightly/173803) I installed this version of the plugin solved this problem, you can try it.

**Reply by LeoMalik** [reply]
@iHx-rainbow @zukxu i have the same issue...downgrade version and set up proxy all didn't help, all i did was just update idea and copilot version.... did you guys fix that problem?

**Reply by LeoMalik** [reply]
@lyh543 seems [1.1.16.1219](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/167546) is the right version for me.

**Reply by KlNon** [reply]
I set up a global proxy in my system environment and it work!

**Reply by 2af** [reply]
Downgrading worked for me on Macbook 2018, thank you!

**Comment by jansorg** [comment]
@lyh543 Thank you for your thorough testing! Are you using a proxy server on your computer or in your IDE? If that's the case, what kind of proxy is it (e.g. HTTP, SOCKS, ...)? Thanks!

**Reply by lyh543** [reply]
Hi, I'm behind a HTTP proxy at localhost. I can test cases with other type of proxies and without proxy after my supper.

**Comment by hsenag** [comment, marked as answer]
Hi, I'm sorry but Copilot does not support proxies at the moment - this is something we intend to remedy in the coming months. I appreciate that older versions sometimes do work with proxies but for technical reasons those methods are no longer supported.

**Reply by lyh543** [reply]
Thank you for you answer. I still wonder if I have to use proxy in IntelliJ, is there any possibility that Copilot can bypass the proxy automatically, instead of throwing errors directly? Or maybe the bypassing proxies is time-consuming to implement?

**Reply by hsenag** [reply]
It doesn't make any explicit effort to connect to proxies at all, but as you identified in your other post, it's possible that libraries it uses do try to respect environment variables etc.

**Reply by neil-rubens** [reply]
@hsenag many major corporations have proxies (for remote work and on-site); so am hoping the fix will come soon
p.s. company laptops have many of the networking settings locked (so proxy workarounds do not work)
p.s.❤️ copilot; and am looking forward to having it back in my daily routine

**Reply by lyh543** [reply]
@neil-rubens If you are able to connect to GitHub directly, you can set `HTTP_PROXY=""` before start IntelliJ IDEA. I can hardly access GitHub without a proxy, since I'm in Mainland China lol

**Reply by neil-rubens** [reply]
@lyh543 thanks a lot for the advice.  Unfortunately disabling proxy doesn't work for the plugin (does work for the JetBrains itself)

**Reply by lyh543** [reply]
@neil-rubens Actually JetBrains reads system proxy, and plugin reads `HTTP_PROXY`. 

> you can set `HTTP_PROXY=""` before start IntelliJ IDEA

By this I mean you can try to:

1. start a terminal
2. run `export HTTP_PROXY=""; export HTTP_PROXY=""` (for Mac OS or Linux) or `$env:HTTP_PROXY = $env:HTTPS_PROXY = ""` (for PowerShell on Windows)
3. in the same terminal, run `/path/to/jetbrains/idea`
4. to validate `HTTP_PROXY` is clear in JetBrains, you can run a terminal in JetBrain, and `echo $HTTP_PROXY` or `echo $env:HTTP_PROXY`

**Reply by hsenag** [reply]
> @hsenag many major corporations have proxies (for remote work and on-site)

@neil-rubens yeah, I'm also quite familiar with this from past roles. We do care about this issue, I'm just not able to make any promises about timing right now.

**Reply by tbenbrahim** [reply]
@meil-rubens, I was getting nagged to upgrade the plugin, so I upgraded and the plugin stopped working I suspect I am not alone. I would expect to see a large decrease in usage from corporate users. Hopefully, this will be fixed before this is sold as a service,

**Reply by neil-rubens** [reply]
@lyh543 thank you very much for the detailed instructions for clearing up the proxy.  I have checked my env proxy value and it was empty; just in case I followed you instructions; and unfortunately results are the same (copilot is not working 😭).

**Reply by lyh543** [reply]
@neil-rubens Oh, that's really strange. Maybe you can try to get the error message in IDEA logs (IDEA titlebar -> `help` -> `show logs in (explorer or other file browser)` -> open the `idea.log` in your file browser).

**Reply by jidaojiuyou** [reply]
Why not provide a way to authorize with auth code(the code from github auth response)

**Reply by fischermatte** [reply]
IMHO this should get high priority. We would only use Copilot in our corporation if it works with idea's proxy settings

**Reply by hantsy** [reply]
Close my proxy, the same result as I described in comment https://github.com/orgs/github-community/discussions/16230#discussioncomment-3020833

**Reply by jerodg** [reply]
I don't see how you will be able to release your "for companies" plan without proxy support. Pretty much everyone coding remotely or in-office for work is going to be running through a proxy. Not sure why this wasn't part of the initial public release. It should be a top priority.

**Comment by lyh543** [comment]
I captured some packages sent from Copilot. It seems that from 1.17 on, Copilot begins to **send requests using `axios`** instead of Java HTTP client.

The good news is that `axios` supports `HTTP_PROXY` and `HTTPS_PROXY` environment by default, which also takes effect in Copilot. The bad news is that, `axios`'s request to HTTPS on HTTP proxy often fails. (https://github.com/axios/axios/issues/3384)

As a GitHub Copilot user, if you want to get Copilot working with proxy on your computer, you can create a demo NodeJS project with `axios` and debug on it. Anyway, I will keep using 1.16.😄

**Reply by hsenag** [reply]
Just FYI: `axios` is actually only one of the libraries we use for outgoing connections. We use some others under different circumstances. And regrettably we'll soon need to withdraw support for the older versions that use the Java HTTP client so that we can make other service upgrades, so I'm afraid your workaround won't work for long.

**Reply by lyh543** [reply]
@hsenag I agree with Copilot team's decision. When 1.16 can no longer working, I will try other workarounds. Thank you again for your technical explanation 👍

**Reply by wuxiaoqiang12** [reply]
Hi, Does 1.16 still work?

**Reply by lyh543** [reply]
> Hi, Does 1.16 still work?

I', afraid it's too old to connect to GitHub Server after 2022.5.20. Now I'm using latest version, with workaround described in the question

**Comment by postrational** [comment]
Unfortunately problems persist with release `1.1.23.1570`:

> Failed to initiate the GitHub login process. Please try again.

**Comment by tbenbrahim** [comment]
@postrational I was able to install [1.1.20.1417](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/172765) manually and it works with a proxy
1.1.23.1570 does not work, and something much older like [1.1.16.1219](https://plugins.jetbrains.com/plugin/17718-github-copilot/versions/stable/167546) also does not work, contrary to what the original poster found. I would note that I have both a proxy defined in IntelliJ settings and the HTTPS_PROXY, NO_PROXY, etc... environment variables set.
This is on Windows 10 running 2022.1.1

It is amazing how quickly one takes AI code completion for granted, I was totally lost this morning trying to write a CloudFormation template without code completion!

**Comment by hantsy** [comment]
Upgrade to 2022.1.3, Github Copilot can not login. 

Click the Copilot icon/Login into Github, the device code login is successful in browser, but IDEA can not receive the authentication feedback result, always showing the waiting progress dialog.

**Reply by CreatorEdition** [reply]
这个是因为 1.20 后的版本不再通过代码其设置的HTTP代理传输数据导致的

**Comment by postrational** [comment]
Latest versions:

* PyCharm `2022.1.4` (Professional Edition) and PyCharm `2022.2` (Professional Edition)
* com.github.copilot (`1.1.28.1744`)

However, I'm still getting the error:

> `Failed to initiate the GitHub login process. Please try again.`

I have correct proxy settings in both the shell and PyCharm settings.
Running on MacOS.

**Comment by postrational** [comment]
Started working for me with the versions below, thank you!

* PyCharm `2022.2.1` (Professional Edition)
* com.github.copilot (`1.1.30.1887`)

**Comment by kalkrishnan** [comment]
IntelliJ IDEA 2022.3.2 and Copilot version 1.2.3.2385 does not work! Get the message please login to GitHub and try again. I am already logged in

**Comment by Shaojay** [comment]
Failed to initiate the GitHub login process. Please try again.  why ?

**Comment by stuaxo** [comment]
@iamcjwilliams can someone on the copilot team look at this ?  Network access in the UI thread is a big no-no, I see this all the time.
